src/misc/unicode.cpp

```diff
@@ -0,0 +1,1159 @@
+        code_page_mapping_reverse_t mapping_file;
+
+        if (!import_mapping_code_page(GetResourcePath(config_mapping.extends_dir),
+                                      config_mapping.extends_file, mapping_file))
+            return false;
+
+        for (const auto &entry : mapping_file)
+             add_to_mappings(entry.first, entry.second);
+    }
+
+    mappings[code_page]         = new_mapping;
+    mappings_reverse[code_page] = new_mapping_reverse;
+    return true;
+}
+
+static void construct_mapping_aliases(const uint16_t code_page)
+{
+    assert(!mappings_aliases.count(code_page));
+    assert(mappings.count(code_page));
+
+    const auto &mapping   = mappings[code_page];
+    auto &mapping_aliases = mappings_aliases[code_page];
+
+    for (const auto &alias : config_aliases)
+        if (!mapping.count(alias.first) && mapping.count(alias.second) &&
+            !mapping_aliases.count(alias.first))
+            mapping_aliases[alias.first] = mapping.find(alias.second)->second;
+}
+
+static bool prepare_code_page(const uint16_t code_page)
+{
+    if (mappings.count(code_page))
+        return true; // code page already prepared
+
+    if (!config_mappings.count(code_page) || !construct_mapping(code_page)) {
+        // Unsupported code page or error
+        mappings.erase(code_page);
+        mappings_reverse.erase(code_page);
+        return false;
+    }
+
+    construct_mapping_aliases(code_page);
+    return true;
+}
+
+static void load_config_if_needed()
+{
+    // If this is the first time we are requested to prepare the code page,
+    // load the top-level configuration and fallback 7-bit ASCII mapping
+
+    static bool config_loaded = false;
+    if (!config_loaded) {
+        const auto path_root = GetResourcePath(dir_name_mapping);
+        import_mapping_ascii(path_root);
+        import_config_main(path_root);
+        config_loaded = true;
+    }
+}
+
+// ***************************************************************************
+// External interface
+// ***************************************************************************
+
+uint16_t UTF8_GetCodePage()
+{
+    const uint16_t cp_default = 437; // United States
+    if (!IS_EGAVGA_ARCH)
+        // Below EGA it wasn't possible to change character set
+        return cp_default;
+
+    load_config_if_needed();
+    const uint16_t cp = deduplicate_code_page(dos.loaded_codepage);
+
+    // For unsupported code pages, revert to default one
+    if (prepare_code_page(cp))
+        return cp;
+    else
+        return cp_default;
+}
+
+void UTF8_RenderForDos(const std::string &str_in, std::string &str_out,
+                       const uint16_t code_page)
+{
+    load_config_if_needed();
+    const uint16_t cp = deduplicate_code_page(code_page);
+    prepare_code_page(cp);
+
+    std::vector<uint16_t> str_wide;
+    utf8_to_wide(str_in, str_wide);
+
+    str_out.clear();
+    wide_to_code_page(str_wide, str_out, cp);
+    str_out.shrink_to_fit();
+}
```

Comment:
If we always clobber the incoming string and then repopulate it, and at the same time aren't using the `void` return value, this is a good candidate to move to a return.
 
Maybe, `std::string UTF8_EncodeforDos(const std::string &str_in,const uint16_t code_page)`

This would make the last bit:

```
    std::string str_out = {};
    wide_to_code_page(str_wide, str_out, cp);
    return str_out;
```

Because `str_out` is a moveable temporary with its lifetime terminating, the compiler will `std::move` its content into the receiving side (even better, I believe it actually constructs the string in-place on the receiving side, called 'return-value-optimization, or RVO: https://stackoverflow.com/a/39750839).

Reply:
OK, changed - but I have kept the `shrink_to_fit()`. AFAIK standard library sometimes allocates a little bit more memory for the string, to make insertions faster by avoiding internal memory reallocation; and we won't be changing the string any more.

Reply:
Very interesting, and good to know.